### ISSUE TITLE
Retry middleware improvements

### ISSUE THREAD

**Issue by LucioFranco** [tower-rs/tower#682]
This issue scopes out the changes we are proposing to the retry middleware to improve its ergonomics. Currently, the retry middleware is quite hard to use and requires implementing a custom `Policy`. Writing this policy is non-trivial and is more work than it should be.

These improvements are aimed at setting up retries with tower to be easier and more user friendly. As well as providing good defaults that work out of the box. 

List of improvements to `tower` and` tower-http`:

- [ ] Simplify `Policy` (https://github.com/tower-rs/tower/pull/681).
  - Change trait fn to take `&mut self`.
  - Change `Future` output to `()`.
  - Alllow `Policy` to be object safe.
- [ ] Add generic backoff utilities.
  - [ ] Add some `trait Backoff` that has an associated future type that allows others to use this utility without being tied to `tokio::time`.
  - [ ] Add a `ExponentialBackoff` type that implements `Backoff`, ported from `linkerd2-proxy`.
- [ ] Budget improvements.
  - [x] Merge budget docs PR (https://github.com/tower-rs/tower/pull/613).
  - [ ] Implement simpler non-time based token bucket budget.
  - [ ] Add some `Budget` trait to allow users to choose which implementation to use.
- [ ] Add a new batteries included standard retry policy
  - [ ] New `StandardRetryPolicy` combining `impl Backoff` and `impl Budget`.
  - [ ] Add `StandardRetryPolicyBuilder` that accept closures (?) for `is_retryable(&mut Req, &mut Result<Res, E>) -> bool` and a `clone_request(&Req) -> Option<Req>`.
- [ ] `tower-http` improvements.
  - [ ] Add new `retry` module
  - [ ] Implement `ReplayBody` similar to the one implemented in `linkerd2-proxy`.
  - [ ] Add new `HttpRetry` layer that accepts higher level constructs for retrying, like `ClassifyResponse`, and will wrap http request bodies with `ReplayBody`.
- [ ] Documentation
  - [ ] Blog post on how to setup retries with `tower` and `tower-http`.
  - [ ] Examples for thick clients with retries in both `tower` and `tower-http`.

## Example code

`tower` examples with no `http`:

```rust
let policy = StandardRetryPolicy::builder()
    .should_retry(|res| match res {
        Ok(_) => false,
        Err(_) => true,
    })
    .clone_request(|r| Some(*r))
    .build();

let mut svc = ServiceBuilder::new()
    .retry(policy)
    .buffer(10)
    .timeout(Duration::from_secs(10))
    .service(svc);
```

`tower-http` examples:

```rust
let make_classifier = ServerErrorsAsFailures::make_classifier();

let mut svc = ServiceBuilder::new()
    .set_request_id("Request-Id".try_into().unwrap(), MakeRequestUuid)
    .retry(StandardHttpPolicy::new(
        make_classifier,
        ExponentialBackoff::default(),
        Budget::default(),
        |e| match e {
            ServerErrorsFailureClass::Status(s) => true,
            ServerErrorsFailureClass::Error(s) => false,
        },
    ))
    .timeout(Duration::from_secs(5))
    .service(client);
```

cc @rcoh @hawkw @jonhoo @seanmonstar @davidpdrsn

**Comment by lovesegfault** [comment]
It would be cool if this could integrate with [429](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429#:~:text=The%20HTTP%20429%20Too%20Many,before%20making%20a%20new%20request.) and the [`Retry-After`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After) header out-of-the-box.

**Comment by LucioFranco** [comment]
@lovesegfault yeah, good idea, this seems like a config option that could live in `HttpRetry`.

**Comment by DontBreakAlex** [comment]
+1 to lovesegfault, I found this crate will looking for a way to handle Retry-After easily

**Comment by xj524598** [comment]
so this issue delay ?
